File: packages/calibration/src/matrix-store.ts
/**
 * W40's current matrix is a union, not a rewritten monolith. Raw row slices travel
 * with the parsed objects so a legacy whole-matrix witness stays checkable without
 * reserialising evidence. Scratch matrices retain the ordinary schema-5 boundary.
 */
import { createHash } from "node:crypto";
import { existsSync, readFileSync } from "node:fs";
import { join, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import { serializeResultCellKey, type CellResult } from "./report";
import { sameFilesystemFile } from "./matrix-write-guard";

export interface MatrixStoreOptions {
  readonly resultsDir?: string;
  readonly matrixPath?: string;
}
export interface Document { readonly path: string; readonly sha256: string }
export interface Entry {
  readonly activeDocumentSha256: string;
  readonly documents: readonly Document[];
  readonly bytes: number;
  readonly sha256: string;
  readonly rowCount: number;
  readonly rowsByProfileKey: Readonly<Record<string, number>>;
  readonly status?: "current" | "retired";
}
export interface Index {
  readonly schemaVersion?: number;
  readonly files: Readonly<Record<string, Entry>>;
  readonly byDocumentSha256: Readonly<Record<string, string | readonly string[]>>;
  readonly sharedReceded?: Readonly<Record<string, readonly string[]>>;
  readonly currentByProfile?: Readonly<Record<string, string>>;
}
interface Owner { readonly file: string; readonly entry: Entry }
const DEFAULT_RESULTS = fileURLToPath(new URL("../results/", import.meta.url));
const rawRows = new WeakMap<CellResult, Buffer>();
const sha = (raw: Buffer) => createHash("sha256").update(raw).digest("hex");
const frozen = (row: CellResult) => row.key.profileKey.startsWith("apple-macos-26.5-");
function fail(message: string): never { throw new Error(`matrix store: ${message}`); }
function sorted(rows: readonly CellResult[]): readonly CellResult[] {
  const keyed = new Map<string, CellResult>();
  for (const row of rows) {
    const key = serializeResultCellKey(row.key);
    if (keyed.has(key)) fail(`duplicate serialized key ${key}`);
    keyed.set(key, row);
  }
  return [...keyed.keys()].sort().map((key) => keyed.get(key)!);
}

/** A brace walk, derived from W30's splitter, preserves each row's original bytes. */
export function readRows(file: string, entry?: Entry): readonly CellResult[] {
  const raw = readFileSync(file);
  if (entry && (raw.length !== entry.bytes || sha(raw) !== entry.sha256)) {
    fail(`${file}: metadata bytes or SHA-256 mismatch`);
  }
  const parsed = JSON.parse(raw.toString("utf8")) as { schemaVersion?: number; cells?: CellResult[] };
  if (parsed.schemaVersion !== 5 || !Array.isArray(parsed.cells)) fail(`${file}: expected schema 5 cells`);
  const cellsKey = raw.indexOf('"cells"');
  let depth = 0, start = 0, rowIndex = 0, quoted = false, escaped = false;
  for (let i = raw.indexOf("[", cellsKey) + 1; i < raw.length; i++) {
    const c = raw[i];
    if (quoted) {
      if (escaped) escaped = false;
      else if (c === 92) escaped = true;
      else if (c === 34) quoted = false;
    } else if (c === 34) quoted = true;
    else if (c === 123 || c === 91) { if (depth++ === 0) start = i; }
    else if (c === 125 || c === 93) {
      if (depth === 0) break;
      if (--depth === 0) {
        const row = parsed.cells[rowIndex++];
        if (!row || typeof row.key?.profileKey !== "string" || typeof row.key.sceneId !== "string") {
          fail(`${file}: malformed row key`);
        }
        rawRows.set(row, raw.subarray(start, i + 1));
      }
    }
  }
  if (rowIndex !== parsed.cells.length) fail(`${file}: raw row span count differs`);
  const rows = sorted(parsed.cells);
  if (entry) validateOwnership(file, rows, entry);
  return rows;
}
export function documents(row: CellResult): { active: Document; receded?: Document } {
  const path = row.key.web.capturePath;
  const matches = [...path.matchAll(/(materialProfile|recededProfile)=(\S+) sha256:([0-9a-f]{12})(?![0-9a-f])/g)];
  const active = matches.filter((m) => m[1] === "materialProfile");
  const receded = matches.filter((m) => m[1] === "recededProfile");
  if (active.length !== 1 || receded.length > 1 || matches.length !== (path.match(/sha256:/g) ?? []).length) {
    fail(`incomplete document clauses in ${path}`);
  }
  return { active: { path: active[0]![2]!, sha256: active[0]![3]! },
    ...(receded[0] ? { receded: { path: receded[0][2]!, sha256: receded[0][3]! } } : {}) };
}
function recededHash(entry: Entry): string | undefined {
  const others = entry.documents.filter((d) => d.sha256 !== entry.activeDocumentSha256);
  if (others.length > 1 || entry.documents.length < 1 || entry.documents.length > 2 ||
      entry.documents.filter((d) => d.sha256 === entry.activeDocumentSha256).length !== 1) {
    fail("generation must name one active and at most one receded document");
  }
  return others[0]?.sha256;
}
function validateOwnership(file: string, rows: readonly CellResult[], entry: Entry) {
  const receded = recededHash(entry);
  const counts: Record<string, number> = {};
  const expected = new Set(entry.documents.map((d) => `${d.path}|${d.sha256}`));
  for (const row of rows) {
    if (frozen(row)) fail(`${file}: frozen row in generation`);
    const named = documents(row);
    if (named.active.sha256 !== entry.activeDocumentSha256 || named.receded?.sha256 !== receded ||
        !expected.has(`${named.active.path}|${named.active.sha256}`) ||
        (named.receded && !expected.has(`${named.receded.path}|${named.receded.sha256}`))) {
      fail(`${file}: row ownership or document clause completeness differs`);
    }
    counts[row.key.profileKey] = (counts[row.key.profileKey] ?? 0) + 1;
  }
  if (rows.length !== entry.rowCount || JSON.stringify(Object.entries(counts).sort()) !==
      JSON.stringify(Object.entries(entry.rowsByProfileKey).sort())) fail(`${file}: row metadata differs`);
}
export function readIndex(results: string, directory: string): Index | undefined {
  const file = join(results, directory, "index.json");
  if (!existsSync(file)) return undefined;
  const index = JSON.parse(readFileSync(file, "utf8")) as Index;
  if (!index.files || !index.byDocumentSha256 ||
      (directory === "generations" && (index.schemaVersion !== 1 || !index.currentByProfile))) {
    fail(`${file}: invalid index schema`);
  }
  const actual = new Map<string, Set<string>>();
  for (const [name, entry] of Object.entries(index.files)) {
    if (!/^[0-9a-f]{12}(?:-[0-9a-f]{12})?\.json$/.test(name)) fail(`${file}: invalid generation filename`);
    recededHash(entry);
    if (directory === "generations" && entry.status !== "current" && entry.status !== "retired") fail(`${file}: invalid status`);
    for (const doc of entry.documents) {
      if (!/^[0-9a-f]{12}$/.test(doc.sha256)) fail(`${file}: invalid document hash`);
      const owners = actual.get(doc.sha256) ?? new Set<string>(); owners.add(name); actual.set(doc.sha256, owners);
    }
  }
  const declared = new Map<string, Set<string>>();
  for (const [hash, names] of Object.entries(index.byDocumentSha256)) {
    declared.set(hash, new Set(typeof names === "string" ? [names] : names));
  }
  for (const [hash, names] of Object.entries(index.sharedReceded ?? {})) {
    const owners = declared.get(hash) ?? new Set<string>();
    for (const name of names) owners.add(name);
    declared.set(hash, owners);
  }
  if (actual.size !== declared.size || [...actual].some(([hash, names]) => {
    const aliases = declared.get(hash);
    return !aliases || names.size !== aliases.size || [...names].some((name) => !aliases.has(name));
  })) fail(`${file}: document aliases do not enumerate every owner`);
  return index;
}
function resultsPath(options: MatrixStoreOptions): string { return resolve(options.resultsDir ?? DEFAULT_RESULTS); }

export function loadCurrentRows(options: MatrixStoreOptions = {}): readonly CellResult[] {
  const results = resultsPath(options);
  const explicit = options.matrixPath ?? process.env["VITREA_MATRIX_PATH"];
  const matrix = join(results, "matrix.json");
  if (explicit && !sameFilesystemFile(resolve(explicit), matrix)) {
    return readRows(resolve(explicit));
  }
  const rows = readRows(matrix);
  const index = readIndex(results, "generations");
  if (!index) return rows;
  if (rows.some((row) => !frozen(row))) fail("non-frozen row in frozen matrix file");
  for (const row of rows) documents(row);
  const selected = index.currentByProfile!;
  const filenames = new Set(Object.values(selected));
  const current: CellResult[] = [...rows];
  for (const [name, entry] of Object.entries(index.files)) {
    const chosen = filenames.has(name);
    if (chosen !== (entry.status === "current")) fail(`${name}: current selection/status differs`);
    if (!chosen) continue;
    for (const profile of Object.keys(entry.rowsByProfileKey)) {
      if (selected[profile] !== name) fail(`${name}: incomplete current profile selection`);
    }
    current.push(...readRows(join(results, "generations", name), entry));
  }
  for (const [profile, name] of Object.entries(selected)) {
    if (!index.files[name]?.rowsByProfileKey[profile]) fail(`${profile}: missing current generation`);
  }
  return sorted(current);
}
export function loadCurrentProfile(profileKey: string, options: MatrixStoreOptions = {}): readonly CellResult[] {
  return loadCurrentRows(options).filter((row) => row.key.profileKey === profileKey);
}
function owners(results: string): Owner[] {
  const out: Owner[] = [];
  for (const directory of ["generations", "superseded"]) {
    const index = readIndex(results, directory);
    for (const [name, entry] of Object.entries(index?.files ?? {})) out.push({ file: join(results, directory, name), entry });
  }
  return out;
}
/** An omitted qualifier resolves an alias; null explicitly selects no receded document. */
export function loadGeneration(active: string, receded?: string | null, options: MatrixStoreOptions = {}): readonly CellResult[] {
  const results = resultsPath(options);
  const found = owners(results).filter(({ entry }) => receded === undefined
    ? entry.documents.some((doc) => doc.sha256 === active)
    : entry.activeDocumentSha256 === active && recededHash(entry) === (receded ?? undefined));
  if (found.length > 1) fail(`ambiguous document ${active}; qualify the active/receded pair`);
  if (found[0]) return readRows(found[0].file, found[0].entry);
  // Frozen generations predate the indexed layout; their documents still name them.
  const frozenRows = readRows(join(results, "matrix.json")).filter((row) => {
    const named = documents(row);
    return named.active.sha256 === active && (receded === undefined || named.receded?.sha256 === (receded ?? undefined));
  });
  const identities = new Set(frozenRows.map((row) => documents(row).receded?.sha256 ?? "none"));
  if (identities.size > 1) fail(`ambiguous document ${active}; qualify the active/receded pair`);
  if (!frozenRows.length) fail(`no generation owns document ${active}`);
  return frozenRows;
}
/** History is opt-in: no current reader pays to load the archive. */
export function* iterateRecordedRows(options: MatrixStoreOptions = {}): IterableIterator<CellResult> {
  const results = resultsPath(options);
  yield* readRows(join(results, "matrix.json"));
  for (const owner of owners(results)) yield* readRows(owner.file, owner.entry);
}
/** Stream the historical pretty-printed envelope, never a regenerated matrix file. */
export function legacyEnvelopeDigest(rows: readonly CellResult[]): string {
  const ordered = sorted(rows);
  const hash = createHash("sha256");
  if (ordered.length === 0) return hash.update('{\n  "schemaVersion": 5,\n  "cells": []\n}\n').digest("hex");
  hash.update('{\n  "schemaVersion": 5,\n  "cells": [\n    ');
  ordered.forEach((row, i) => {
    const raw = rawRows.get(row);
    if (!raw) fail("legacy envelope requires rows loaded with their raw bytes");
    if (i) hash.update(",\n    ");
    hash.update(raw);
  });
  hash.update("\n  ]\n}\n");
  return hash.digest("hex");
}

/** Preserve recorded row slices in the splitter's original, key-ordered envelope. */
export function generationEnvelope(rows: readonly CellResult[]): Buffer {
  const ordered = sorted(rows);
  if (!ordered.length) return Buffer.from('{\n  "schemaVersion": 5,\n  "cells": []\n}\n');
  return Buffer.concat([Buffer.from('{\n  "schemaVersion": 5,\n  "cells": [\n    '),
    ...ordered.flatMap((row, i) => {
      const raw = rawRows.get(row);
      if (!raw) fail("generation envelope requires raw row bytes");
      return i ? [Buffer.from(",\n    "), raw] : [raw];
    }), Buffer.from("\n  ]\n}\n")]);
}

File: packages/calibration/results/2026-09-26-w40-g0-generations/matrix_store.py
#!/usr/bin/env python3.12
"""Python read-only adapter for W40's frozen bed and indexed generations.

Cells retain their original raw byte slices privately, so legacy-envelope hashes
certify recorded evidence rather than a Python reserialization of its numbers.
"""
import collections
import hashlib
import json
import os
import pathlib
import re

from snapshot import RESULTS, digest, documents, key, slices

_RAW = {}
_ENVELOPE = {}
_TOKEN = re.compile(r"^[0-9a-f]{12}$")
_FILE = re.compile(r"^[0-9a-f]{12}(?:-[0-9a-f]{12})?\.json$")
_UNQUALIFIED = object()


def _load_matrix(path):
    raw = path.read_bytes()
    envelope = json.loads(raw)
    if envelope.get("schemaVersion") != 5 or not isinstance(envelope.get("cells"), list):
        raise ValueError(f"{path}: expected schema-5 cells array")
    spans, prefix, sep, suffix = slices(raw)
    if len(spans) != len(envelope["cells"]):
        raise ValueError(f"{path}: cells and raw byte slices differ")
    rows = []
    seen = set()
    for (a, b), parsed in zip(spans, envelope["cells"]):
        row = json.loads(raw[a:b])
        if row != parsed:
            raise ValueError(f"{path}: parsed row differs from byte slice")
        identity = key(row)
        if identity in seen:
            raise ValueError(f"{path}: duplicate serialized row key {identity}")
        seen.add(identity)
        _RAW[id(row)] = raw[a:b]
        _ENVELOPE[id(row)] = (prefix, sep, suffix)
        rows.append(row)
    return rows, raw


def _index(directory, archived=False):
    path = directory / "index.json"
    index = json.loads(path.read_bytes())
    if not archived and (index.get("schemaVersion") != 1 or
                         not all(isinstance(index.get(x), dict) for x in
                                 ("files", "byDocumentSha256", "currentByProfile"))):
        raise ValueError(f"{path}: invalid generations index schema")
    if archived and not all(isinstance(index.get(x), dict) for x in
                            ("files", "byDocumentSha256")):
        raise ValueError(f"{path}: invalid superseded index")
    owners = collections.defaultdict(set)
    for filename, metadata in index["files"].items():
        if not _FILE.fullmatch(filename):
            raise ValueError(f"{path}: invalid file name {filename}")
        active = metadata.get("activeDocumentSha256")
        docs = metadata.get("documents")
        if not _TOKEN.fullmatch(active or "") or not isinstance(docs, list):
            raise ValueError(f"{path}: invalid active hash/documents in {filename}")
        hashes = [d.get("sha256") for d in docs]
        if (len(hashes) not in (1, 2) or active not in hashes or len(set(hashes)) != len(hashes)
                or any(not isinstance(d.get("path"), str) or not _TOKEN.fullmatch(h or "")
                       for d, h in zip(docs, hashes))):
            raise ValueError(f"{path}: incomplete document pair in {filename}")
        if not archived and (metadata.get("status") not in ("current", "retired") or
                             not isinstance(metadata.get("rowsByProfileKey"), dict)):
            raise ValueError(f"{path}: invalid generation entry {filename}")
        for h in hashes:
            owners[h].add(filename)
    if not archived:
        declared = index["byDocumentSha256"]
        if any(not isinstance(names, list) or len(names) != len(set(names))
               for names in declared.values()) or {
                   h: sorted(names) for h, names in declared.items()
               } != {h: sorted(names) for h, names in owners.items()}:
            raise ValueError(f"{path}: incomplete or stale aliases")
    if archived:
        archive_aliases = {h: sorted({first, *index.get("sharedReceded", {}).get(h, [])})
                           for h, first in index["byDocumentSha256"].items()}
        if archive_aliases != {h: sorted(names) for h, names in owners.items()}:
            raise ValueError(f"{path}: incomplete archive aliases")
    return index


def _file(directory, filename, metadata):
    path = directory / filename
    raw = path.read_bytes()
    if len(raw) != metadata.get("bytes") or digest(raw) != metadata.get("sha256"):
        raise ValueError(f"{path}: byte length or SHA-256 differs from index")
    rows, _ = _load_matrix(path)
    if len(rows) != metadata.get("rowCount"):
        raise ValueError(f"{path}: row count differs from index")
    expected_docs = {(d["path"], d["sha256"]) for d in metadata["documents"]}
    counts = collections.Counter()
    for row in rows:
        if row["key"]["profileKey"].startswith("apple-macos-26.5-"):
            raise ValueError(f"{path}: frozen row in indexed generation")
        clauses = documents(row)
        if {(p, h) for _, p, h in clauses} != expected_docs:
            raise ValueError(f"{path}: row document ownership differs from index")
        if next(h for kind, _, h in clauses if kind == "materialProfile") != metadata["activeDocumentSha256"]:
            raise ValueError(f"{path}: row active document differs from index")
        counts[row["key"]["profileKey"]] += 1
    if dict(counts) != metadata.get("rowsByProfileKey"):
        raise ValueError(f"{path}: profile counts differ from index")
    return rows


def _unique(rows):
    seen = set()
    for row in rows:
        identity = key(row)
        if identity in seen:
            raise ValueError(f"serialized row key occurs in two authoritative files: {identity}")
        seen.add(identity)
    return sorted(rows, key=lambda row: key(row))


def _current(results_dir):
    frozen_path = results_dir / "matrix.json"
    frozen, _ = _load_matrix(frozen_path)
    index_path = results_dir / "generations/index.json"
    if not index_path.exists():
        return _unique(frozen), None
    for row in frozen:
        if not row["key"]["profileKey"].startswith("apple-macos-26.5-"):
            raise ValueError(f"{frozen_path}: indexed layout must contain frozen profiles only")
        documents(row)
    directory = index_path.parent
    index = _index(directory)
    members = {}
    for name, entry in index["files"].items():
        if entry["status"] == "current":
            members[name] = _file(directory, name, entry)
    for profile, name in index["currentByProfile"].items():
        if name not in members or index["files"][name]["status"] != "current" or not any(
            row["key"]["profileKey"] == profile for row in members[name]
        ):
            raise ValueError(f"{index_path}: missing current profile {profile} in {name}")
    for name, entry in index["files"].items():
        for profile in entry["rowsByProfileKey"]:
            if entry["status"] == "current" and index["currentByProfile"].get(profile) != name:
                raise ValueError(f"{index_path}: current file {name} not selected for {profile}")
            if entry["status"] == "retired" and index["currentByProfile"].get(profile) == name:
                raise ValueError(f"{index_path}: retired file selected for {profile}")
    current = frozen + [row for name in sorted(set(index["currentByProfile"].values())) for row in members[name]]
    return _unique(current), (index, members)


def load_current_rows(matrix_path=None, results_dir=None):
    """Key-ordered current union; explicit paths and VITREA_MATRIX_PATH are scratch files."""
    override = matrix_path or os.environ.get("VITREA_MATRIX_PATH")
    directory = pathlib.Path(results_dir) if results_dir else RESULTS
    if override and pathlib.Path(override).resolve() != (directory / "matrix.json").resolve():
        return _unique(_load_matrix(pathlib.Path(override))[0])
    return _current(directory)[0]


def load_current_profile(profile_key, results_dir=None):
    rows = load_current_rows(results_dir=results_dir)
    selected = [row for row in rows if row["key"]["profileKey"] == profile_key]
    return selected


def owners_for_document(sha256, results_dir=None):
    """Enumerate EVERY indexed current, retired and archive owner of a document."""
    directory = pathlib.Path(results_dir) if results_dir else RESULTS
    owners = []
    for subdir, archived in (("generations", False), ("superseded", True)):
        path = directory / subdir
        if not (path / "index.json").exists():
            continue
        index = _index(path, archived)
        if archived:
            names = {index["byDocumentSha256"][sha256],
                     *index.get("sharedReceded", {}).get(sha256, [])} if sha256 in index["byDocumentSha256"] else set()
        else:
            names = set(index["byDocumentSha256"].get(sha256, []))
        owners.extend((subdir, name) for name in sorted(names))
    return owners


def load_generation(active, receded=_UNQUALIFIED, results_dir=None):
    """Resolve the document pair; explicit None selects the no-receded generation."""
    if not _TOKEN.fullmatch(active) or (receded is not _UNQUALIFIED and receded is not None
                                        and not _TOKEN.fullmatch(receded)):
        raise ValueError("generation documents require twelve-hex SHA-256 tokens")
    directory = pathlib.Path(results_dir) if results_dir else RESULTS
    choices = []
    for subdir, name in owners_for_document(active, directory):
        index = _index(directory / subdir, subdir == "superseded")
        entry = index["files"][name]
        if receded is not _UNQUALIFIED and entry["activeDocumentSha256"] != active:
            continue
        secondary = next((d["sha256"] for d in entry["documents"]
                          if d["sha256"] != entry["activeDocumentSha256"]), None)
        if receded is _UNQUALIFIED or receded == secondary:
            choices.append((subdir, name, entry))
    if len(choices) > 1:
        qualifier = "unspecified" if receded is _UNQUALIFIED else "none" if receded is None else receded
        raise ValueError(f"generation ({active}, {qualifier}) has "
                         f"{len(choices)} owners; supply an unambiguous active/receded pair")
    if choices:
        subdir, name, entry = choices[0]
        return sorted(_file(directory / subdir, name, entry), key=key)
    # Frozen macOS 26.5 generations predate the indexed layout.
    frozen, _ = _load_matrix(directory / "matrix.json")
    matching = [row for row in frozen if any(
        kind == "materialProfile" and h == active for kind, _, h in documents(row)
    ) and (receded is _UNQUALIFIED or (
        next((h for kind, _, h in documents(row) if kind == "recededProfile"), None) == receded
    ))]
    secondaries = {next((h for kind, _, h in documents(row) if kind == "recededProfile"), None)
                   for row in matching}
    if len(secondaries) > 1:
        raise ValueError(f"ambiguous frozen generation {active}; supply receded qualifier")
    if not matching:
        raise ValueError(f"no generation owns document {active}")
    return sorted(matching, key=key)


def iterate_recorded_rows(results_dir=None):
    """Opt-in iterator across current, retired and historical evidence."""
    directory = pathlib.Path(results_dir) if results_dir else RESULTS
    yield from load_current_rows(results_dir=directory)
    for subdir, archived in (("generations", False), ("superseded", True)):
        path = directory / subdir
        if not (path / "index.json").exists():
            continue
        index = _index(path, archived)
        for name, entry in index["files"].items():
            if archived or entry["status"] == "retired":
                yield from _file(path, name, entry)


def legacy_envelope_digest(rows):
    """Streaming digest over the ORIGINAL row bytes and schema-5 envelope."""
    ordered = _unique(rows)
    if not ordered:
        return digest(b'{\n  "schemaVersion": 5,\n  "cells": []\n}\n')
    sha = hashlib.sha256()
    sha.update(b'{\n  "schemaVersion": 5,\n  "cells": [\n    ')
    for i, row in enumerate(ordered):
        if i:
            sha.update(b",\n    ")
        if id(row) not in _RAW:
            raise ValueError("legacy envelope requires rows loaded with their raw bytes")
        sha.update(_RAW[id(row)])
    sha.update(b"\n  ]\n}\n")
    return sha.hexdigest()
